### PR TITLE
Pytest with 89% coverage

### PR DESCRIPTION
**Pull request by rflamary**
+ Add numerous test for existing functions and classes.
+ Correct failing build due to Python3/2 map function difference.

Will merge soon since currently POT do not build.

### CODE REVIEW

**Comment by rflamary** [comment]
POT now build with a certain number of tests. I will merge this PR today unless somebody objects.

File: .travis.yml [hunk 1]
```diff
@@ -13,6 +13,10 @@ matrix:
           python: 2.7
 before_install:
   - ./.travis/before_install.sh
+before_script: # configure a headless display to test plot generation
+  - "export DISPLAY=:99.0"
+  - "sh -e /etc/init.d/xvfb start"
+  - sleep 3 # give xvfb some time to start
```
Comment (agramfort) [inline .travis.yml]: do you use only matplotlib? if so just use the Agg backend

File: Makefile [hunk 2]
```diff
@@ -38,10 +38,10 @@ pep8 :
 	flake8 examples/ ot/ test/
 
 test : FORCE pep8
-	python -m py.test -v test/
+	python -m py.test -v test/ --cov=ot --cov-report html:cov_html
 	
 pytest : FORCE 
-	python -m py.test -v test/	
+	python -m py.test -v test/ --cov=ot
```
Comment (agramfort) [inline Makefile]: you should have a native pytest command:

pytest -v test/ --cov=ot

File: test/test_bregman.py [hunk 3]
```diff
@@ -37,7 +66,68 @@ def test_sinkhorn_variants():
...
+    M /= M.max()
+
+    alpha = 0.5  # 0<=alpha<=1
+    weights = np.array([1 - alpha, alpha])
+
+    # wasserstein
+    reg = 1e-3
+    bary_wass = ot.bregman.barycenter(A, M, reg, weights)
+
+    assert np.allclose(1, np.sum(bary_wass))
```
Comment (agramfort) [inline test/test_bregman.py]: you have an assert_allclose function in numpy

**Comment by agramfort** [comment]
@rflamary please wait. I'll do a proper review in the next 2 days.

File: .travis.yml (hunk 1 above)
Comment (rflamary) [inline .travis.yml]: I tried to use the Agg backedn here
https://github.com/rflamary/POT/blob/pytest/test/test_plot.py

but travis still failed with DISPLAY error
https://travis-ci.org/rflamary/POT/builds/256924206

maybe the test_plot.py is not good but I don't see the problem

File: Makefile (hunk 2 above)
Comment (rflamary) [inline Makefile]: Under Debian/Ubuntu logilab-common install a useless executable named pytest. It's a well known bug buttakes time to be corrested. This line ensure that the proper py.test is executed.

https://github.com/pytest-dev/pytest/issues/1833

**Comment by agramfort** [comment]
ok you win

sklearn is not as careful

https://github.com/scikit-learn/scikit-learn/blob/master/build_tools/travis/test_script.sh#L25

File: .travis.yml (hunk 1 above)
Comment (agramfort) [inline .travis.yml]: that's weird. You seem to have done it right. Are you sure matplotlib is  not imported anywhere before?

you should also nest the imports to matplotlib in the or source tree. So matplotlib is not imported when you do `import ot`

File: test/test_bregman.py [hunk 4]
```diff
@@ -23,6 +23,35 @@ def test_sinkhorn():
     assert np.allclose(u, G.sum(0), atol=1e-05)  # cf convergence sinkhorn
 
 
+def test_sinkhorn_empty():
+    # test sinkhorn
+    n = 100
+    np.random.seed(0)
```
Comment (agramfort) [inline test/test_bregman.py]: use a random state

rng = np.random.RandomState(42)
x = rng.randn(n, 2)

etc.

ie don't change the global seed.

File: test/test_bregman.py [hunk 5]
```diff
@@ -23,6 +23,35 @@ def test_sinkhorn():
...
+
+    x = np.random.randn(n, 2)
+    u = ot.utils.unif(n)
+
+    M = ot.dist(x, x)
+
+    G, log = ot.sinkhorn([], [], M, 1, stopThr=1e-10, verbose=True, log=True)
+    # check constratints
+    assert np.allclose(u, G.sum(1), atol=1e-05)  # cf convergence sinkhorn
+    assert np.allclose(u, G.sum(0), atol=1e-05)  # cf convergence sinkhorn
```
Comment (agramfort) [inline test/test_bregman.py]: use np.testing.assert_allclose

it makes errors clearer than just an assert

File: test/test_bregman.py [hunk 6]
```diff
@@ -37,7 +66,68 @@ def test_sinkhorn_variants():
...
+    assert np.allclose(G0, Gerr)
+
+
+def test_bary():
+
+    n = 100  # nb bins
+
+    # Gaussian distributions
+    a1 = ot.datasets.get_1D_gauss(n, m=30, s=10)  # m= mean, s= std
+    a2 = ot.datasets.get_1D_gauss(n, m=40, s=10)
```
Comment (agramfort) [inline test/test_bregman.py]: as I was saying it should be named in the future

make_1d_gauss

File: test/test_bregman.py [hunk 7]
```diff
@@ -37,7 +66,68 @@ def test_sinkhorn_variants():
...
+    bary_wass = ot.bregman.barycenter(A, M, reg, weights)
+
+    assert np.allclose(1, np.sum(bary_wass))
+
+    ot.bregman.barycenter(A, M, reg, log=True, verbose=True)
+
+
+def test_unmix():
+
+    n = 50  # nb bins
```
Comment (agramfort) [inline test/test_bregman.py]: n -> n_bins

as n can mean n_samples etc.

if you call it n_bins no need to write nb bins :)

File: test/test_da.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+
+
+import ot
+import numpy as np
```
Comment (agramfort) [inline test/test_da.py]: import numpy before ot
as ot depends on numpy

it's for convention

File: test/test_da.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+# import pytest
+
+
+def test_OTDA():
```
Comment (agramfort) [inline test/test_da.py]: test_OTDA -> test_otda

no caps in function names

File: test/test_da.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    n = 150  # nb bins
+
+    xs, ys = ot.datasets.get_data_classif('3gauss', n)
+    xt, yt = ot.datasets.get_data_classif('3gauss2', n)
```
Comment (agramfort) [inline test/test_da.py]: get_data_classif -> make_classification

would be sklearn consistent.

File: test/test_dr.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
+import ot
+import numpy as np
+import pytest
+
+try:  # test if cudamat installed
+    import ot.dr
```
Comment (agramfort) [inline test/test_dr.py]: test for what you really need to test ie if cudamat is available

```
try:
    import cudamat
    has_cudamat = False
except ...:
    has_cudamat = True
```

File: test/test_dr.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    nogo = False
+except ImportError:
+    nogo = True
+
+
+@pytest.mark.skipif(nogo, reason="Missing modules (autograd or pymanopt)")
```
Comment (agramfort) [inline test/test_dr.py]: now it says autograd and pymanopt :)

File: test/test_dr.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+def test_fda():
+
+    n = 90  # nb samples in source and target datasets
+    np.random.seed(0)
```
Comment (agramfort) [inline test/test_dr.py]: use RandomState

File: test/test_da.py (hunk 10 above)
Comment (rflamary) [inline test/test_da.py]: OK we should open an issue and handle that in a separate PR I think, this one is mainly for testing

File: test/test_dr.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+    # generate circle dataset
+    xs, ys = ot.datasets.get_data_classif('gaussrot', n)
+
+    nbnoise = 8
```
Comment (agramfort) [inline test/test_dr.py]: n_features_noise

File: test/test_dr.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+    projfda(xs)
+
+    assert np.allclose(np.sum(Pfda**2, 0), np.ones(p))
+
+
+@pytest.mark.skipif(nogo, reason="Missing modules (autograd or pymanopt)")
+def test_wda():
+
+    n = 100  # nb samples in source and target datasets
```
Comment (agramfort) [inline test/test_dr.py]: n -> n_samples

File: test/test_dr.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    nz = 0.2
+    np.random.seed(0)
```
Comment (agramfort) [inline test/test_dr.py]: RandomState

File: test/test_dr.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        (np.cos(t).reshape((-1, 1)), np.sin(t).reshape((-1, 1))), 1)
+    xs = xs * ys.reshape(-1, 1) + nz * np.random.randn(n, 2)
+
+    nbnoise = 8
+
+    xs = np.hstack((xs, np.random.randn(n, nbnoise)))
+
+    p = 2
+
+    Pwda, projwda = ot.dr.wda(xs, ys, p, maxiter=10)
```
Comment (agramfort) [inline test/test_dr.py]: if you step back a little guessing for a new user what `ot.dr.wda` means it's a pretty hard challenge :)

maybe you can leaving with a bit more typing and use less jardon/acronyms

File: test/test_gpu.py [hunk 18]
```diff
@@ -3,16 +3,23 @@
...
-    import ot.gpu
+
+    np.random.seed(0)
 
     def describeRes(r):
         print("min:{:.3E}, max::{:.3E}, mean::{:.3E}, std::{:.3E}".format(
             np.min(r), np.max(r), np.mean(r), np.std(r)))
 
-    for n in [5000]:
+    for n in [50, 100, 500, 1000]:
```
Comment (agramfort) [inline test/test_gpu.py]: for n_samples ...

File: test/test_gpu.py [hunk 19]
```diff
@@ -3,16 +3,23 @@
...
+except ImportError:
+    nogpu = True
+
 
-@pytest.mark.skip(reason="No way to test GPU on travis yet")
+@pytest.mark.skipif(nogpu, reason="No GPU available")
 def test_gpu_sinkhorn():
-    import ot.gpu
+
+    np.random.seed(0)
```
Comment (agramfort) [inline test/test_gpu.py]: RandomState

File: test/test_gpu.py [hunk 20]
```diff
@@ -30,30 +37,36 @@ def describeRes(r):
...
-            time3 = time.time()
-            print("Normal sinkhorn lpl1, time: {:6.2f} sec ".format(
-                time2 - time1))
-            describeRes(G1)
-            print("   GPU sinkhorn lpl1, time: {:6.2f} sec ".format(
-                time3 - time2))
-            describeRes(G2)
+    np.random.seed(0)
+
+    def describeRes(r):
```
Comment (agramfort) [inline test/test_gpu.py]: describeRes -> describe_result

no CamelCase in functions

File: test/test_optim.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+import ot
+import numpy as np
+
+# import pytest
+
+
+def test_conditional_gradient():
+
+    n = 100  # nb bins
```
Comment (agramfort) [inline test/test_optim.py]: n_bins

File: test/test_optim.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
+
+
+import ot
+import numpy as np
```
Comment (agramfort) [inline test/test_optim.py]: numpy first

**Comment by agramfort** [comment]
Also please add license and authors on top of each file.

**Comment by rflamary** [comment]
@agramfort  Do I copy the full MIT or do I use a short statement with 
# Author:
# MIT

as in sklearn ?

Thank you for the review I will implement most of your comments and open Issues for the function renaming thing.3

When using get_dataset functions I have to use random.seed because those functions use the random.randn generator.

File: test/test_dr.py (hunk 12 above)
Comment (rflamary) [inline test/test_dr.py]: corrected the comment top of the test file.

**Comment by rflamary** [comment]
OK @agramfort  I took care of most your reviews.

What remains and will be opened as Issues :
- Renaming dataset function to be more sklearn compliant (breaking change)
- Weird travis fail with no open DISPLAY

If the travis build work I will merge the PR since I introduced no features in the toolbox only tests.

File: test/test_da.py [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+# License: MIT License
+
+import numpy as np
+import ot
+
+
+def test_otda():
+
+    n_samples = 150  # nb samples
+    np.random.seed(0)
```
Comment (agramfort) [inline test/test_da.py]: RandomState

the get_data_classif function should take the rng in param and use it instead of np.random.randn

see the check_random_state function in sklearn

File: test/test_optim.py [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+# License: MIT License
+
+import numpy as np
+import ot
+
+
+def test_conditional_gradient():
+
+    n_bins = 100  # nb bins
+    np.random.seed(0)
```
Comment (agramfort) [inline test/test_optim.py]: RandomState

File: test/test_utils.py [hunk 25]
```diff
@@ -15,9 +18,9 @@ def f(i):
 
     a = np.arange(n)
 
-    l1 = map(f, a)
+    l1 = list(map(f, a))
 
-    l2 = ot.utils.parmap(f, a)
+    l2 = list(ot.utils.parmap(f, a))
 
     assert np.allclose(l1, l2)
```
Comment (agramfort) [inline test/test_utils.py]: use np.testing.assert_allclose

File: test/test_utils.py [hunk 26]
```diff
@@ -71,6 +74,52 @@ def test_dist():
             D[i, j] = np.sum(np.square(x[i, :] - x[j, :]))
 
     D2 = ot.dist(x, x)
+    D3 = ot.dist(x)
 
     # dist shoul return squared euclidean
     assert np.allclose(D, D2)
+    assert np.allclose(D, D3)
```
Comment (agramfort) [inline test/test_utils.py]: idem

File: test/test_utils.py [hunk 27]
```diff
@@ -71,6 +74,52 @@ def test_dist():
...
+
+
+def test_dist0():
+
+    n = 100
+    M = ot.utils.dist0(n, method='lin_square')
+
+    # dist0 default to linear sampling with quadratic loss
+    assert np.allclose(M[0, -1], (n - 1) * (n - 1))
```
Comment (agramfort) [inline test/test_utils.py]: idem and below too

**Comment by agramfort** [comment]
ok no more nitpicks after these

**Comment by rflamary** [comment]
OK great thank you again,

I won't handle the rng stuff in this PR I will add it to the Issue about the make_datasets function.

**Comment by agramfort** [comment]
k

**Comment by agramfort** [comment]
+1 for merge

**Comment by rflamary** [comment]
OK let's merge this PR,

We now have a 89% coverage of the code when all libraries are installed (cudamat, pymanopt, autograd).

Also the Makefile include the test target that checks for PEP8 violations and and run the tests. 

I have created Issue #20 for the dataset function names and random state problems.
